### ISSUE TITLE
Bug, Firestore enablePresistence not completed or fail to catch while offline

### ISSUE THREAD

**Issue by mohshraim** [firebase/firebase-js-sdk#687]
### [REQUIRED] Describe your environment

  * Operating System version: mac , chrome
  * Firebase SDK version: 4.12.1
  * Firebase Product:  firestore

### [REQUIRED] Describe the problem

firebase.firestore().enablePresistence(). then().catch()
neither completed nor failed to catch while offline

#### Steps to reproduce:

1- load the web page
2- turn wifi off
3- hit button that init the firestore with presistance

#### Relevant Code:

    function start(){
        firebase.firestore().enablePresistence(). then( function(){
              console.log('init done');
        }).catch(function(err){
              console.log('init fail', err);
       });
    }

Iam not sure if this behavior is expected;  even if i turn the wifi-on after click the init button..

**Comment by mikelehen** [comment]
That's certainly not expected.  Are you sure start() is getting called? 😁 

Our tests / sample code don't hit this issue so in order to investigate we'll need a way to reproduce it. Would it be possible to create a jsfiddle / jsbin / whatever that demonstrates the issue?

**Comment by mohshraim** [comment]
Hello @mikelehen

Seems the problem related to auth sdk..  and only after the first success login with some data loading..

i can't build jsfiddle; that jsfiddle will not enable you to refresh page while you are offline... this case requiere to load firebase JS SDK files from local; iam using webserver chrome...

it took time to make it simple to minimize code case

html body have three buttons (init firebase,  signin, startDB )

1- load page and click on initFirebsae  then  signin   then  StartDB.... 
--> everything should work normally on while you are online
2- go offline refresh the page, if you click on  initFirebsae then startDB -->   here is the problem 

 ==> you will notice that initFirebase that listen to  firebase.auth().onAuthStateChanged ; it seems that firebase.auth().onAuthStateChanged fail to get any update on this case then enablePresistence not working at all...


just copy the following code and load js SDK files on your directory...


Thanks

        ` <input type="button" onclick="initFirebase();" value="init firebase" >
        
        <input type="button" onclick="mySignIn();" value="signIn" >

       <input type="button" onclick="startDB();" value="startDB" >


       
     
     <script src="assets/fire/firebase.js"></script>
     <script src="assets/fire/firebase-app.js"></script>
     <script src="assets/fire/firebase-auth.js"></script>
     <script src="assets/fire/firebase-firestore.js"></script>

     <script>
          var initFirebase = function () {
               // Initialize Firebase
               var config = {
                    apiKey: "",
                    authDomain: "firebaseapp.com",
                    databaseURL: ".com",
                    projectId: "",
                    storageBucket: ".com",
                    messagingSenderId: ""
               };
               firebase.initializeApp(config);

               firebase.auth().onAuthStateChanged(function (user) {
                    // logOnNative('onAuthStateChanged => ' + JSON.stringify(user));
                    console.log('onAuthStateChanged => ', user);
               });

          };


          var fdb;

          var initFirestoreDB = function () {
               return new Promise(function (resolve, reject) {
                    firebase.firestore().enablePersistence()
                         .then(function () {
                              // Initialize Cloud Firestore through firebase
                              fdb = firebase.firestore();
                              console.log('enablePersistence done');
                              resolve({
                                   ok: true
                              });
                         })
                         .catch(function (err) {
                              console.log('firestore enablePersistence error:' + err);
                              reject(err);
                         });
               });
          };

          var mySignIn = function (storeUsername, storePassword) {

               firebase.auth().signInAndRetrieveDataWithEmailAndPassword('moh.shraim@hotmail.com','passssss')
               .then(function (result) {
                    console.log(result);
               })
               .catch(function (error) {
                    console.log(error);
               });
          };

          function startDB() {

               initFirestoreDB().then(function (result) {

                         fdb.collection("category")
                              .onSnapshot(function (querySnapshot) {
                                   querySnapshot.forEach(function (doc) {
                                        console.log(doc.data());
                                   });
                              });
                    })
                    .catch(function (error) {
                         console.log(error);
                    });
          }
     </script>`

**Comment by mohshraim** [comment]
Hello @mikelehen

should i reference this error with new issue to firebase authentication?? or if you want to change the label to component-auth ?
 
it seems that  `firebase.initializeApp(config); `  will fail on offline state (wifi-off) if user signed-in in any previous session; 
and this will block  `firebase.firestore().enablePersistence()` from returning any result..

thanks for help

**Comment by mohshraim** [comment]
any help. please!

**Comment by mikelehen** [comment]
@bojeil-google Is there somebody from auth that could look at this?  It seems that onAuthStateChanged() isn't firing at all when offline and this breaks Firestore (technically we listen for app.INTERNAL.addAuthTokenListener() but it's likely the same underlying issue?)

**Comment by mohshraim** [comment]
@bojeil-google 

please check repo 

https://github.com/mohshraim/MCVE_firebase_issue_687

**Comment by mohshraim** [comment]
@bojeil-google 

you should make a successful login before turn then turn wifi-off and refresh the page. and init..

Please check this video
https://github.com/mohshraim/MCVE_firebase_issue_687/blob/master/fire_auth_bug.mov

on the other hand @mikelehen also confirm this

Please note on repo i use v.4.12   i cant find v.4.12.1 at 
https://github.com/firebase/firebase-bower/releases

Thanks for follow up.

**Comment by mikelehen** [comment]
Thanks @mohshraim for the details. Just to be clear, I haven't tried to reproduce this myself, so I can't confirm it at this point.

**Comment by mohshraim** [comment]
My Apology..
it seems that i didnt update the Auth.js file when i update the other one to v.4.12
sorry for your time..
